File: py-algorithm/number/haming_distance.py
import unittest

from math import gcd
from fractions import Fraction
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/hamming-distance/

    Đếm số khác nhau tương tứng của 2 số nguyên trong biểu diễn nhị phân

        1   (0 0 0 1)
        4   (0 1 0 0)
              ↑   ↑     -> Kết quả = 2
"""


class Solution(unittest.TestCase):

    def hammingDistance(self, x: int, y: int) -> int:
        b1 = format(x, 'b')
        b2 = format(y, 'b')
        len1, len2 = len(b1), len(b2)
        max_len = max(len1, len2)
        count = 0
        for i in range(max_len):
            c1 = b1[- 1 - i] if i < len1 else '0'
            c2 = b2[- 1 - i] if i < len2 else '0'
            if c1 != c2: count += 1

        return count

    def test(self):
        self.assertEqual(2, self.hammingDistance(1, 4))
        self.assertEqual(1, self.hammingDistance(1, 3))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/number/HamingDistance.java
package number;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/*-
 * https://leetcode.com/problems/hamming-distance/
 *
 * Đếm số khác nhau tương tứng của 2 số nguyên trong biểu diễn nhị phân
 *
 *  1   (0 0 0 1)
 *  4   (0 1 0 0)
 *         ↑   ↑     -> Kết quả = 2
 */
public class HamingDistance {

    public int hammingDistance(int x, int y) {
        String b1 = Integer.toString(x, 2);
        String b2 = Integer.toString(y, 2);
        int count = 0;
        int len1 = b1.length(), len2 = b2.length();
        int max = Math.max(len1, len2);
        for (int i = 0; i < max; i++) {
            char c1 = i < len1 ? b1.charAt(len1 - 1 - i) : '0';
            char c2 = i < len2 ? b2.charAt(len2 - 1 - i) : '0';
            if (c1 != c2) count++;
        }
        return count;
    }

    @Test
    public void test() {
        assertEquals(2, hammingDistance(1, 4));
        assertEquals(1, hammingDistance(1, 3));
    }
}
